### PR TITLE
Add session store and middleware for session handling

### PR DESCRIPTION
**Pull request by franciscog-kainos**
** Design **
The API exposes a singleton service which contains the `SessionStore` and the session handler `RequestHandler` for express.js projects.

The design is based on the [Railway Oriented Design](https://medium.com/@naveenkumarmuguda/railway-oriented-programming-a-powerful-functional-programming-pattern-ab454e467f31)

* Purify-ts was used to enhance the expressibility of the functions in the project. 
* `ErrorFunctions.ts` contains functions which are used to handle exceptional (`Left`) cases. This should help reason more about the logic and less about the implementation.

### CODE REVIEW

File: src/SessionMiddlewareFactory.ts [hunk 1]
```diff
@@ -1,37 +1,65 @@
...
+                    (verifiedSession: VerifiedSession) => {
+                        request.session = verifiedSession;
+                    }
+                );
 
             } else {
-                /**
-                 * Generate and redirect?
-                 */
+                request.session = VerifiedSession.createNewVerifiedSession(this.config)
```
Comment (geek-soft) [inline src/SessionMiddlewareFactory.ts]: Why does middleware create session?

File: src/AuthenticationMiddlewareFactory.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { Request, Response, NextFunction } from "express";
+import { SessionKeys } from "./session/SessionKeys";
+
+export type RedirectFunction = (response: Response) => void;
+export const createAuthenticationMiddleware = (redirectFn: RedirectFunction) =>
```
Comment (geek-soft) [inline src/AuthenticationMiddlewareFactory.ts]: Makes sense to remove it since we said we will keep it session focused.

File: src/index.ts [hunk 3]
```diff
@@ -1,5 +1,50 @@
...
+class CHSessionServiceImpl implements CHSessionService {
+
+    private static impl: CHSessionServiceImpl;
+    private sessionMiddlewareFactory: SessionMiddlewareFactory;
+
+    public readonly sessionStore: SessionStore;
+    public readonly sessionHandler: RequestHandler;
+
+    private constructor(public readonly config: SessionHandlerConfig) {
+        const redis = Cache.redisInstance(config);
```
Comment (geek-soft) [inline src/index.ts]: This is super confusing. Cache class is used in two ways:
1) to produce redis instance
2) to wrap redis instance

This hides real purpose of the class. Is it more focused on hiding implementation details (raw redis client) or is it not?

File: src/index.ts [hunk 4]
```diff
@@ -1,5 +1,50 @@
...
+import { SessionMiddlewareFactory } from "./SessionMiddlewareFactory";
+import { SessionHandlerConfig } from "./SessionHandlerConfig";
+import { SessionStore } from "./session/SessionStore";
+import { Cache } from "./cache/Cache";
+import { RequestHandler } from "express";
 
-import SessionMiddlewareFactory from "./SessionMiddlewareFactory";
 
-export = SessionMiddlewareFactory;
+export interface CHSessionService {
```
Comment (geek-soft) [inline src/index.ts]: What is the purpose of this service? Does it have one clearly defined purpose of existence?

File: src/SessionMiddlewareFactory.ts [hunk 5]
```diff
@@ -1,37 +1,66 @@
...
+export class SessionMiddlewareFactory {
 
-    static create(): RequestHandler {
+    constructor(private readonly config: SessionHandlerConfig,
+        private readonly sessionStore: SessionStore,
+    ) {
 
-        return async function (request, _response, next) {
+        if (!config.cookieSecret) {
+            throw Error("Must provide secret of at least 16 bytes (64 characters) long");
```
Comment (geek-soft) [inline src/SessionMiddlewareFactory.ts]: This error message and one below (at line 31) mismatch. Shall that be 64 characters or 24? Misleading error messages are equally bad as lack of error message.

File: src/SessionMiddlewareFactory.ts [hunk 6]
```diff
@@ -1,37 +1,66 @@
...
+            throw Error("Must provide expiry period");
+        }
+
+        if (!config.cacheServer) {
+            throw Error("Must provide redis url");
+        }
+
+        if (config.cookieSecret.length < 24) {
+            console.log(config.cookieSecret.length);
+            throw Error("Secret must be at least 16 bytes (24 characters) long");
```
Comment (geek-soft) [inline src/SessionMiddlewareFactory.ts]: How 24 characters maps to 16 bytes?

File: src/session/SessionStore.ts [hunk 7]
```diff
@@ -1,109 +1,66 @@
...
-    }
+    };
 
-    private decodeSession(encodedData: any) {
+    public deleteSession = (cookieString: string): EitherAsync<Failure, number> => {
+        return liftToAsyncEither<Failure, string>(cookieString).chain(this.cache.del);
+    };
 
-        const base64Decoded = Encoding.decodeBase64(encodedData);
+    public load = <T>(cookieId: string): EitherAsync<Failure, T> => {
```
Comment (geek-soft) [inline src/session/SessionStore.ts]: Why are load and store public? What is their relationship with `loadSession` and `storeSession` functions?

File: src/session/SessionStore.ts [hunk 8]
```diff
@@ -1,109 +1,66 @@
...
+import { Cookie } from "./model/Cookie";
+
+export class SessionStore {
+
+
+    public constructor(private readonly cache: Cache) {
     }
 
-    private getAccessTokenData(data: any) {
+    private getAccessTokenData(session: Session): AccessToken | undefined {
```
Comment (geek-soft) [inline src/session/SessionStore.ts]: It is never used, looks like candidate for removal.

File: src/session/SessionStore.ts [hunk 9]
```diff
@@ -1,109 +1,66 @@
...
-
-        const accessTokenData = this.getAccessTokenData(data);
-
-        if (accessTokenData && accessTokenData.expiresIn) {
-            this.validateExpiration(accessTokenData.expiresIn);
-        }
-
-        return new Session(sessionId, accessTokenData);
-    }
+    public loadSession = (sessionSecret: string, cookie: string): EitherAsync<Failure, VerifiedSession> => {
```
Comment (geek-soft) [inline src/session/SessionStore.ts]: I feel that function that loads session should not be responsible for cookie verification. It is one more responsibility too many for the method.

File: src/session/model/Session.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    private constructor(session: Session) {
+        super();
+        this.data = session.data;
+    }
+
+    public asCookie = (): Cookie => {
+        return Cookie.sessionCookie(this);
+    };
+
+    public static createNewVerifiedSession(
```
Comment (geek-soft) [inline src/session/model/Session.ts]: At this point I see no use case for it. It also looks wrong to issue 'verified' session while access tokens are dummy. Isn't accounts component the only one entitled to session creation?

File: src/session/model/Session.ts [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        sessionData[SessionKeys.ClientSig] = newCookie.signature;
+        sessionData[SessionKeys.SignInInfo] = signInInfo;
+        sessionData[SessionKeys.Expires] = Date.now() + accessToken.expires_in*1000;
+
+        return new VerifiedSession(new Session(sessionData));
+    }
+
+    public static verifySession(session: Session): Either<Failure, VerifiedSession> {
+
+        const signInInfo = session.data[SessionKeys.SignInInfo];
```
Comment (geek-soft) [inline src/session/model/Session.ts]: Is that even possible to have session without 'sign in' info? Can access token be missing when accounts created session? Can there be session without expires?

File: src/cache/cache.ts [hunk 12]
```diff
@@ -1,28 +1,58 @@
...
+import { Failure } from "../error/FailureType";
+import { PromiseError, NoDataRetrievedError } from "../error/ErrorFunctions";
+import { eitherPromiseToEitherAsync } from "../utils/EitherUtils";
+import { Cookie as string } from "../session/model/Cookie";
+import { SessionHandlerConfig } from "../SessionHandlerConfig";
+import { Redis } from 'ioredis';
+import IORedis = require('ioredis');
 
-import config from "../config";
+export class Cache {
```
Comment (n-r-b) [inline src/cache/cache.ts]: I think 'Cache' is a bit misleading here as we are using Redis for session storage and not as a cache

Comment (n-r-b) [reply src/cache/cache.ts]: Also the class name has an uppercase 'C' yet the filename is lowercase, we should be consistent here

File: src/cache/cache.ts [hunk 13]
```diff
@@ -1,28 +1,58 @@
...
+        };
+
+        const promise = () => this.client.get(key)
+            .then(checkIfResultEmpty)
+            .catch(err => Left<Failure, string>(Failure(PromiseError(err))));
+
+        return wrapPromise(promise());
+    };
+
+    public del = (key: string): EitherAsync<Failure, number> => {
```
Comment (n-r-b) [inline src/cache/cache.ts]: Are we deleting or expiring sessions in redis?

File: src/SessionMiddlewareFactory.ts (hunk 6 above)
Comment (franciscog-kainos) [reply src/SessionMiddlewareFactory.ts]: Should say base64 encoded string

File: src/session/SessionMiddleware.ts [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+import expressAsyncHandler from "express-async-handler";
+
+
+export function SessionMiddleware(config: CookieConfig, sessionStore: SessionStore): RequestHandler {
+    return initializeRequestHandler(config, sessionStore);
+}
+
+function initializeRequestHandler(config: CookieConfig, store: SessionStore): RequestHandler {
+
+    if (!config.cookieSecret) {
```
Comment (n-r-b) [inline src/session/SessionMiddleware.ts]: can we combine these two if statements? Error is identical

Comment (n-r-b) [reply src/session/SessionMiddleware.ts]: do we need to check there's a cookie name too?

File: src/session/model/Cookie.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+        return new Cookie(session.data[SessionKey.Id], session.data[SessionKey.ClientSig]);
+    };
+
+    public static newCookie(sessionSecret: string): Cookie {
+        const sessionId = generateSessionId();
+        const signature = generateSignature(sessionId, sessionSecret);
+        return new Cookie(sessionId, signature);
+    }
+
+    public static validateCookieString = (sessionSecret: string) => (cookieString: string): Either<Failure, Cookie> => {
```
Comment (n-r-b) [inline src/session/model/Cookie.ts]: I don't feel this validate method belongs here in the Cookie model

Comment (n-r-b) [reply src/session/model/Cookie.ts]: Can we also make the naming consistent? we are passing in a `cookieSecret` yet this method accepts a `sessionSecret`

**Pull request merged**